source/extensions/filters/common/expr/evaluator.cc

```diff
@@ -138,14 +143,66 @@ BuilderInstanceSharedPtr createBuilder(Protobuf::Arena* arena) {
     throw CelException(absl::StrCat("failed to register extension regex functions: ",
                                     ext_register_status.message()));
   }
-  return std::make_shared<BuilderInstance>(std::move(builder));
+  // Register string extension functions only if enabled in configuration.
+  if (enable_string_functions) {
+    auto string_register_status =
+        cel::extensions::RegisterStringsFunctions(builder->GetRegistry(), options);
+    if (!string_register_status.ok()) {
+      throw CelException(absl::StrCat("failed to register extension string functions: ",
+                                      string_register_status.message()));
+    }
+  }
+  return builder;
 }
 
-SINGLETON_MANAGER_REGISTRATION(expression_builder);
+// Cache to store builders for different configurations.
+class BuilderCache : public Singleton::Instance {
```

Comment:
I think the singleton needs to be stored in the compiled expression or it will be deleted together with its cache?